src/main/java/io/supertokens/passwordless/Passwordless.java

```diff
@@ -124,6 +127,56 @@ private static String generateUserInputCode() {
         return sb.toString();
     }
 
+    public static DeviceWithCodes getDeviceWithCodesById(Main main, String deviceId)
+            throws StorageQueryException, StorageTransactionLogicException, NoSuchAlgorithmException {
+        byte[] deviceIdBytes = Base64.decodeBase64(deviceId);
+        String deviceIdHash = Base64.encodeBase64URLSafeString(Utils.hashSHA256Bytes(deviceIdBytes));
```

Comment:
Perhaps you want to move this to a helper function since you convert from deviceId (string) -> deviceIdHash (string) in several places.